pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,831 @@
+	assert_eq!(session.sandbox().balance_of(&TOKEN, &ALICE), value);
+	// Successfully emit event.
+	assert_eq!(
+		last_contract_event(&session).unwrap(),
+		Transfer { from: None, to: Some(account_id_from_slice(ALICE.as_ref())), value }
+			.encode()
+			.as_slice()
+	);
+
+	// Total supply increased by `value` exceeds maximal value of `u128` type.
+	assert_eq!(mint(&mut session, ALICE, u128::MAX), Err(into_psp22_custom(Arithmetic(Overflow))));
+
+	// Token is not live, i.e. frozen or being destroyed.
+	assert_ok!(session.sandbox().start_destroy(&TOKEN));
+	// `pallet-assets` returns `AssetNotLive` error.
+	assert_eq!(
+		mint(&mut session, ALICE, value),
+		Err(into_psp22_custom(Module { index: 52, error: [16, 0] }))
+	);
+}
+
+/// 4. PSP-22 Burnable Interface:
+/// - burn
+
+#[drink::test(sandbox = Sandbox)]
+fn burn_works(mut session: Session) {
+	let _ = env_logger::try_init();
+	// No permission to burn.
+	assert_ok!(session.sandbox().create(&(TOKEN + 1), &BOB, MIN_BALANCE)); // Create a new token owned by `BOB`.
+	assert_ok!(session.sandbox().mint_into(&(TOKEN + 1), &BOB, AMOUNT));
+	assert_ok!(deploy(&mut session, CONTRACT.clone(), "existing", vec![(TOKEN + 1).to_string()]));
+	// `pallet-assets` returns `NoPermission` error.
+	assert_eq!(
+		burn(&mut session, BOB, AMOUNT),
+		Err(into_psp22_custom(Module { index: 52, error: [2, 0] }))
+	);
+
+	// Deploy a new contract.
+	let contract = deploy(
+		&mut session,
+		CONTRACT.clone(),
+		"new",
+		vec![TOKEN.to_string(), MIN_BALANCE.to_string()],
+	)
+	.unwrap();
+	let value = 1;
+	session.set_actor(contract.clone());
+	// Mint tokens.
+	assert_ok!(session.sandbox().mint_into(&TOKEN, &ALICE, AMOUNT));
+
+	// No-op.
+	assert_ok!(burn(&mut session, ALICE, 0));
+	assert_eq!(last_contract_event(&session), None); // No event emitted.
+
+	// Failed with `InsufficientBalance`.
+	assert_eq!(burn(&mut session, ALICE, AMOUNT * 2), Err(PSP22Error::InsufficientBalance));
+
+	// Successfully burn tokens.
+	assert_ok!(burn(&mut session, ALICE, value));
+	assert_eq!(session.sandbox().total_supply(&TOKEN), AMOUNT - value);
+	assert_eq!(session.sandbox().balance_of(&TOKEN, &ALICE), AMOUNT - value);
+	// Successfully emit event.
+	assert_eq!(
+		last_contract_event(&session).unwrap(),
+		Transfer { from: Some(account_id_from_slice(ALICE.as_ref())), to: None, value }
+			.encode()
+			.as_slice()
+	);
+
+	// Token is not live, i.e. frozen or being destroyed.
+	assert_ok!(session.sandbox().start_destroy(&TOKEN));
+	// `pallet-assets` returns `IncorrectStatus` error.
+	assert_eq!(
+		burn(&mut session, ALICE, value),
+		Err(into_psp22_custom(Module { index: 52, error: [17, 0] }))
+	);
+}
+
+/// A set of helper methods to test the contract calls.
+mod utils {
+	use drink::session::{error::SessionError, NO_SALT};
+	use pop_api::primitives::AccountId;
+	use pop_sandbox::{AccountId32, Balance, INIT_VALUE};
+	use scale::Decode;
+
+	use super::*;
+
+	/// Deploy test utilities.
+
+	pub(super) fn deploy(
```

Comment:
If possible, should be moved to pop-drink as generic functions taking account id and error.  Defining this in every example, or contract, adds complexity and developer burden.

Reply:
Done